### ISSUE TITLE
thread '<unnamed>' panicked at 'cannot access a TLS value during or after it is destroyed: AccessError'

### ISSUE THREAD

**Issue by rmanus** [nokia/memory-profiler#12]
On debian10 with nigthly rust

```
thread '<unnamed>' panicked at 'cannot access a TLS value during or after it is destroyed: AccessError', src/libcore/result.rs:999:5
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:71
             at src/libstd/sys_common/backtrace.rs:59
             at src/libstd/panicking.rs:197
   2: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:211
             at src/libstd/panicking.rs:474
   3: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:381
   4: rust_begin_unwind
             at src/libstd/panicking.rs:308
   5: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
   6: core::result::unwrap_failed
             at /rustc/37ff5d388f8c004ca248adb635f1cc84d347eda0/src/libcore/macros.rs:18
   7: memory_profiler::unwind::grab
             at /root/.cargo/git/checkouts/not-perf-e01bfa01482c86ed/9739e8b/nwind/src/local_unwinding.rs:0
             at preload/src/unwind.rs:166
   8: calloc
             at preload/src/lib.rs:1423
             at preload/src/lib.rs:1470
   9: g_malloc0
  10: g_slice_free_chain_with_offset
  11: g_queue_free
  12: __nptl_deallocate_tsd.part.8
  13: start_thread
  14: clone
Aborted (core dumped)
```

**Comment by koute** [comment]
Thanks for the report.

What application are you trying to profile? I'd be easier to fix this if I'd have access to a reproduction.

**Comment by rmanus** [comment]
Unfortunately, I can't share the application. But I suspect it has something to do with libpackagekit-glib2-dev.

**Comment by rmanus** [comment]
I was able to reproduce the issue with a simple test case :

```
# cat testcase.c
#include <packagekit-glib2/packagekit.h>
#include <stdio.h>
#include <pthread.h>

void *print_repo_data(void *);

int main(int argc, char **argv) {
	pthread_t thr;
	pthread_attr_t attr;

	pthread_attr_init(&attr);
	pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_DETACHED);
	pthread_create(&thr, &attr, &print_repo_data, NULL);

	return 0;
}

void *print_repo_data(void *data) {
	g_autoptr(GError) gerror      = NULL;
	g_autoptr(PkError) error_code = NULL;
	g_autoptr(PkResults) results  = NULL;
	GPtrArray *array              = NULL;
	PkTask *task                  = NULL;
	PkRepoDetail *item            = NULL;
    	int i;

	task = pk_task_new();

	results = pk_task_get_repo_list_sync(task, PK_FILTER_ENUM_NONE, NULL, NULL, NULL, &gerror);

	/* get the packages returned */
	array = pk_results_get_repo_detail_array(results);
	for (i = 0; i < array->len; i++) {
		item = g_ptr_array_index(array, i);

		printf("%s %s\n", pk_repo_detail_get_id(item), pk_repo_detail_get_description(item));
	}


	g_object_unref(task);
	g_ptr_array_unref(array);
}
```

`# gcc -o testcase -I/usr/include/PackageKit -DI_KNOW_THE_PACKAGEKIT_GLIB2_API_IS_SUBJECT_TO_CHANGE -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/glib-2.0 -lpackagekit-glib2 -lgobject-2.0 -lglib-2.0 -lpthread testcase.c`

```
# LD_PRELOAD=./libmemory_profiler.so ./testcase
...
thread '<unnamed>' panicked at 'cannot access a TLS value during or after it is destroyed: AccessError', src/libcore/result.rs:999:5
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
Aborted (core dumped)
```

This is only happening when pk_* functions are called from a thread. pk_task_get_repo_list_sync() will create a glib pool and create another thread to execute in this pool. The coredump is happening when freeing metadata for this thread.

**Comment by koute** [comment]
Thanks a lot for the reproduction!

I still need to add a test for this, but it should be fixed on `master`!

**Comment by rmanus** [comment]
Thanks - I confirm it is fixing the issue in my code as well.

**Comment by koute** [comment]
Released as 0.2.0.

Thanks again for the report and the reproduction!
